### DISCUSSION TITLE
Adding custom properties from Discord

### DISCUSSION THREAD

**Discussion by ariscript** [nextauthjs/next-auth#1696] [Help]
I am trying to authenticate users with Discord using the "identify" and "guilds" scopes. It seems to be possible to add the user info provided by Discord into the token into `token.profile` (note I am also including the guilds array there as well) using this code:
```ts
async jwt(token, user, account, profile, isNewUser) {
    if (profile) token.profile = profile;
    if (account?.accessToken) token.accessToken = account.accessToken;

    const data = await (
        await fetch("https://discord.com/api/v8/users/@me/guilds", {
            method: "GET",
            headers: {
                Authorization: `Bearer ${account?.accessToken}`,
            },
        })
    ).json();

    // @ts-ignore
    token.profile.guilds = data;

    return token;
},
```
But, as soon as I add a `session` function that adds the `profile` object into the session, the session becomes `null` all the time even when signed in, and the login screen from Discord prompts me to enter username and password instead of just automatically letting me in.
```ts
// @ts-ignore
async session(session, token) {
    console.log("token", token);
    console.log("session", session);

    // @ts-ignore
    session.profile = token.profile;

    return session;
},
```
*As a side-note, the reason I have to add so many `@ts-ignore`s in the `session` function is that it expects a `Promise<WithAdditionalParams<Session>>` while I only return a `Promise<Session>`, not sure how to fix that.*

Can anyone help me add the information I receive from Discord into the `session` so that I can use it in my application?

**Comment by balazsorban44** [comment]
You should do all the logic in the same if block, otherwise things will be overridden. Check the docs: https://next-auth.js.org/configuration/callbacks#jwt-callback

`user, account, profile, isNewUser` are only available at signIn. You can check any of them, all the others will also exist actually. Try:

```js
export default NextAuth({
  callbacks: {
    async jwt(token, user, account, profile, isNewUser) {
      // Sign-in
      if (user) {
        token.profile = user;
        token.accessToken = account.accessToken;
        const res = await fetch("https://discord.com/api/v8/users/@me/guilds", {
          headers: { Authorization: `Bearer ${account.accessToken}` },
        })
        const data = await res.json()
        token.profile.guilds = data
      }
      return token
    },
    session(session, token) {
      session.profile = token.profile
      return session;
    },
  }
  providers: [
    Providers.Discord({
      clientId: process.env.CLIENT_ID,
      clientSecret: process.env.CLIENT_SECRET,
      profile(profile) {
        // What you return here will be the same as the second argument in the jwt() callback at signin.
        return
      }
    })
  ]
})
```

The fourth parameter is actually ALL the information the provider will return, you probably don't need everything. You can filter it down with the provider's profile callback.

Regarding TS, better typings are coming built-in. Follow #1649

**Reply by ariscript** [reply]
The `jwt` callback seems to work fine based on it logging my user info, but the `session` callback seems to never get called. It never logs anything I put there, and such it is always null. The `token` object that goes out of the jwt callback seems to be correct though. As soon as I add a session callback, I get sent to a different login screen funnily (Discord doesn't automatically pick up that I am logged in and asks for me to log in again) and the session is still null.

**Reply by balazsorban44** [reply]
I'm sorry, I may have posted a bit confusing config. Make sure that you have both `session` and `jwt` in the callbacks object. (I updated my code above!)

There is a session property on the root of the config, but it is something different. I wonder if you accidentally used that to define the session callback: https://next-auth.js.org/configuration/options#session

You will want: https://next-auth.js.org/configuration/callbacks#session-callback

The `session` callback is fired right after `jwt`, so it would be strange if you couldn't even log there.:

https://github.com/nextauthjs/next-auth/blob/1c662e9ddc5700dbfe5e945e94ef6246598acfa7/src/server/routes/session.js#L42-L43

**Reply by ariscript** [reply]
I tried literally copying the code above, adding my required scopes and a log in the session callback. It still does not log anything.
```ts
import NextAuth from "next-auth";
import Providers from "next-auth/providers";

export default NextAuth({
    callbacks: {
        async jwt(token, user, account, profile, isNewUser) {
            // Sign-in
            if (user) {
                token.profile = user;
                token.accessToken = account.accessToken;
                const res = await fetch(
                    "https://discord.com/api/v8/users/@me/guilds",
                    {
                        headers: {
                            Authorization: `Bearer ${account.accessToken}`,
                        },
                    }
                );
                const data = await res.json();
                // @ts-ignore
                token.profile.guilds = data;
            }
            console.log(token);
            return token;
        },
        // @ts-ignore
        session(session, token) {
            console.log("session cb");
            // @ts-ignore
            session.profile = token.profile;
            return session;
        },
    },
    providers: [
        Providers.Discord({
            clientId: process.env.DISCORD_ID,
            clientSecret: process.env.DISCORD_SECRET,
            scope: "identify guilds",
            // @ts-ignore
            profile(profile) {
                return profile;
            },
        }),
    ],
});
```
I still feel like something to do with Discord showing me a different login screen might be the issue as before that, the login worked just fine (also coincided with creating a session callback, so I'm not sure).

**Reply by ariscript** [reply]
Okay, I figured something out! The code has nothing wrong with it **until I fetch the guilds**, adapting some code from #1629, I am able to get the user's profile, just not the guilds. As soon as I add that request to Discord's API, the whole thing breaks. Any ideas as to why that might be the case? It would be much more convenient if I can just keep the user's guilds inside the session rather than having to make a request every time I need to use it.

**Reply by ariscript** [reply]
I made this work by fetching and setting guilds in the `session` callback as opposed to `jwt`. I feel like the documentation for this could get a good rework after 2 days of headscratching.

**Reply by balazsorban44** [reply]
you can fetch extra information in the profile callback. The second param will contain access token. The profile callback is only invoked on signin

You definitely don't want to fetch data in the session callback. it will contact discord on each request for the session endpoint

https://next-auth.js.org/configuration/providers#using-a-custom-provider


You could try wrapping your async call in a try..catch when fetching guild, and see if the problem happens there

**Reply by ariscript** [reply]
I tried to put the request in the profile function, and the session immediately became null. In the token, the profile object seems to be correct, but once again, it `session` seems to never get called.

**Reply by balazsorban44** [reply]
and you don't get an error in your console either?

When the fetch call is either in jwt or profile, do you get to log it right after the res.json call? if not it would indicate a failed request and would explain the problem try to wrap it in a try..catch

**Reply by ariscript** [reply]
Yes, the request doesn't fail and logs properly, but nothing in session ever runs. Here is my code for reference: https://sourceb.in/HHhY2G8n0h
